beamer/events.py

```diff
@@ -334,13 +334,14 @@ def _fetch_range(
                 event_abis=self._event_abis,
             )
 
-    def fetch(self) -> list[Event]:
-        try:
-            block_data = self._web3.eth.get_block("latest")
-        except requests.exceptions.ConnectionError:
-            raise
-        except RequestException:
-            return []
+    def fetch(self, block_data=None) -> list[Event]:
```

Comment:
Why does `fetch` need to receive `block_data`?

Reply:
`block_data=None` indicates it doesn't need but it can receive it. When block data is provided, `fetch()` doesn't need to call `self._web3.eth.get_block("latest")` and go straight to fetching logs. 

`fetch()` might be called from two context:
* during the sync or http fallback
* after a block was received by websocket

In the latter, block data is available and can be provided to `fetch()` and above applies.

Reply:
> `block_data=None` indicates it doesn't need but it can receive it. When block data is provided, `fetch()` doesn't need to call `self._web3.eth.get_block("latest")` and go straight to fetching logs.
> 
> `fetch()` might be called from two context:
> 
>     * during the sync or http fallback
> 
>     * after a block was received by websocket
> 
> 
> In the latter, block data is available and can be provided to `fetch()` and above applies.

What I'm trying to understand is, why does `fetch` need to receive the entire block data?
Do you just need the "target block number"?

Reply:
It could be done this way but then this line must be changed what makes the change bigger:
https://github.com/beamer-bridge/beamer/blob/main/beamer/events.py#L345

I think the change follows the original implementation of the function which indeed requests a full block and has access to it. If the function was about to always need only block number, we can refactor so it calls a function that returns a block number instead of block data.